frontend/catalyst/api_extensions/control_flow.py

```diff
@@ -512,36 +518,80 @@ def _call_with_quantum_ctx(self, ctx):
-        args, args_tree = tree_flatten([])
-        args_avals = tuple(map(_abstractify, args))
-        branch_jaxprs, consts, out_treedefs = initial_style_jaxprs_with_common_consts1(
-            (*self.branch_fns, self.otherwise_fn), args_tree, args_avals, "cond"
+                    res_classical_tracers = [inner_trace.full_raise(t) for t in wfun.call_wrapped()]
+
+            regions.append(
+                HybridOpRegion(
+                    inner_trace,
+                    quantum_tape,
+                    [],
+                    collapse(out_sig.out_type(), res_classical_tracers),
+                )
+            )
+            in_sigs.append(in_sig)
+            out_sigs.append(out_sig)
+
+        _assert_cond_result_structure([s.out_tree() for s in out_sigs])
+        _assert_cond_result_types([[t[0] for t in s.out_type()] for s in out_sigs])
+        out_tree = out_sigs[-1].out_tree()
+        all_consts = [s.out_consts() for s in out_sigs]
+        out_types = [s.out_type() for s in out_sigs]
+        # TODO: Creating output tracers like this is not quite cocrect. We need to calculate the
+        # unified result type first and only then use it as the output type of this primitive.
+        out_type2 = out_types[-1]
+
+        out_expanded_classical_tracers = output_type_to_tracers(
+            out_type2,
+            sum(all_consts, []),
+            (),
+            maker=lambda aval: new_inner_tracer(outer_trace, aval),
         )
-        out_signatures = [jaxpr.out_avals for jaxpr in branch_jaxprs]
-        _assert_cond_result_structure(out_treedefs)
-        _assert_cond_result_types(out_signatures)
-        branch_jaxprs = unify_jaxpr_result_types(branch_jaxprs)
-        out_classical_tracers = cond_p.bind(*(self.preds + consts), branch_jaxprs=branch_jaxprs)
-        return tree_unflatten(out_treedefs[0], out_classical_tracers)
+
+        out_classical_tracers = collapse(out_type2, out_expanded_classical_tracers)
+
+        self._operation = Cond(
+            in_classical_tracers,
+            out_classical_tracers,
+            regions,
+            expansion_strategy=self.expansion_strategy,
+        )
+        return tree_unflatten(out_tree, out_classical_tracers)
+
+    def _call_with_classical_ctx(self, ctx):
+        in_classical_tracers = self.preds
+
+        def _trace(branch_fn):
+            _, in_sig, out_sig = trace_function(
+                ctx, branch_fn, *(), expansion_strategy=cond_expansion_strategy()
+            )
+            return in_sig, out_sig
+
+        in_sigs, out_sigs = unzip2(_trace(fun) for fun in (*self.branch_fns, self.otherwise_fn))
+        _assert_cond_result_structure([s.out_tree() for s in out_sigs])
+        _assert_cond_result_types([[t[0] for t in s.out_type()] for s in out_sigs])
+        all_jaxprs = [s.out_initial_jaxpr() for s in out_sigs]
+        all_consts = [s.out_consts() for s in out_sigs]
+        num_implicit_outputs = out_sigs[-1].num_implicit_outputs()
```

Comment:
What is in `out_sigs[-1]`?

Reply:
It was the signature of the last branch. I have removed the magic constant by passing all values to the unification function as-is.